File: CoffeeMaker.ml
(* Define the CoffeeMaker module *)
module CoffeeMaker = struct
  (* Define the functions *)
  let boil_water () =
    print_endline "Boiling water...";
    Ok "Water is boiled"

  let grind_beans () =
    print_endline "Grinding beans...";
    Ok "Beans are ground"

  let make_coffee (water_status, beans_status) =
    match (water_status, beans_status) with
    | (Ok "Water is boiled", Ok "Beans are ground") ->
        print_endline "Making coffee...";
        Ok "Coffee is ready!"
    | _ ->
        Error "Can't make coffee until water is boiled and beans are ground."
end

(* Functional pipeline *)
let () =
  let water_status = CoffeeMaker.boil_water () in
  let beans_status = CoffeeMaker.grind_beans () in
  let coffee_status = CoffeeMaker.make_coffee (water_status, beans_status) in
  match coffee_status with
  | Ok message -> print_endline message
  | Error message -> print_endline message

File: CoffeeMaker.hs
-- Define the Main module
module Main where

import Control.Monad (void)

-- Define the CoffeeMaker functions
boilWater :: IO (Either String String)
boilWater = do
  putStrLn "Boiling water..."
  return (Right "Water is boiled")

grindBeans :: IO (Either String String)
grindBeans = do
  putStrLn "Grinding beans..."
  return (Right "Beans are ground")

makeCoffee :: (Either String String, Either String String) -> IO (Either String String)
makeCoffee (Right "Water is boiled", Right "Beans are ground") = do
  putStrLn "Making coffee..."
  return (Right "Coffee is ready!")
makeCoffee _ = return (Left "Can't make coffee until water is boiled and beans are ground.")

-- Functional pipeline
main :: IO ()
main = do
  waterStatus <- boilWater
  beansStatus <- grindBeans
  coffeeStatus <- makeCoffee (waterStatus, beansStatus)
  case coffeeStatus of
    Right message -> putStrLn message
    Left message -> putStrLn message
